### PR TITLE
[RateLimiter] handle error results of DateTime::modify()

### PR DESCRIPTION
**Pull request by xabbuh**
| Q             | A
| ------------- | ---
| Branch?       | 5.4
| Bug fix?      | yes
| New feature?  | no
| Deprecations? | no
| Issues        | 
| License       | MIT

Depending on the version of PHP the `modify()` method will either throw an exception or issue a warning.

### CODE REVIEW

File: src/Symfony/Component/RateLimiter/RateLimiterFactory.php [hunk 1]
```diff
@@ -68,19 +68,21 @@ public function create(?string $key = null): LimiterInterface
...
         $intervalNormalizer = static function (Options $options, string $interval): \DateInterval {
-            try {
-                // Create DateTimeImmutable from unix timesatmp, so the default timezone is ignored and we don't need to
-                // deal with quirks happening when modifying dates using a timezone with DST.
-                $now = \DateTimeImmutable::createFromFormat('U', time());
+            // Create DateTimeImmutable from unix timesatmp, so the default timezone is ignored and we don't need to
+            // deal with quirks happening when modifying dates using a timezone with DST.
+            $now = \DateTimeImmutable::createFromFormat('U', time());
 
-                return $now->diff($now->modify('+'.$interval));
```
Comment (xabbuh) [inline src/Symfony/Component/RateLimiter/RateLimiterFactory.php:76]: `modify()` returns `false` on failure, thus we must not pass it to `diff()` before checking the type

File: src/Symfony/Component/RateLimiter/RateLimiterFactory.php [hunk 2]
```diff
@@ -68,19 +68,21 @@ public function create(?string $key = null): LimiterInterface
...
-            } catch (\Exception $e) {
-                if (!preg_match('/Failed to parse time string \(\+([^)]+)\)/', $e->getMessage(), $m)) {
-                    throw $e;
-                }
+            try {
+                $nowPlusInterval = @$now->modify('+' . $interval);
```
Comment (xabbuh) [inline src/Symfony/Component/RateLimiter/RateLimiterFactory.php:76]: up to PHP 8.2 the method emitted a warning on invalid intervals which we need to silence here

File: src/Symfony/Component/RateLimiter/RateLimiterFactory.php [hunk 3]
```diff
@@ -68,19 +68,21 @@ public function create(?string $key = null): LimiterInterface
...
+            } catch (\DateMalformedStringException $e) {
```
Comment (xabbuh) [inline src/Symfony/Component/RateLimiter/RateLimiterFactory.php:77]: from PHP 8.3 on a `DateMalformedStringException` is thrown instead

File: src/Symfony/Component/RateLimiter/RateLimiterFactory.php [hunk 4]
```diff
@@ -68,19 +68,21 @@ public function create(?string $key = null): LimiterInterface
...
-
-                return $now->diff($now->modify('+'.$interval));
-            } catch (\Exception $e) {
-                if (!preg_match('/Failed to parse time string \(\+([^)]+)\)/', $e->getMessage(), $m)) {
-                    throw $e;
-                }
+                $nowPlusInterval = @$now->modify('+' . $interval);
 
-                throw new \LogicException(sprintf('Cannot parse interval "%s", please use a valid unit as described on https://www.php.net/datetime.formats.relative.', $m[1]));
+                if (!$nowPlusInterval) {
```
Comment (xabbuh) [inline src/Symfony/Component/RateLimiter/RateLimiterFactory.php:78]: `modify()` returns `false` on failure, thus we must not pass it to `diff()` before checking the type

File: src/Symfony/Component/RateLimiter/RateLimiterFactory.php [hunk 5]
```diff
@@ -68,19 +68,21 @@ public function create(?string $key = null): LimiterInterface
...
 
-                return $now->diff($now->modify('+'.$interval));
-            } catch (\Exception $e) {
-                if (!preg_match('/Failed to parse time string \(\+([^)]+)\)/', $e->getMessage(), $m)) {
-                    throw $e;
-                }
+            try {
+                $nowPlusInterval = @$now->modify('+' . $interval);
+            } catch (\DateMalformedStringException $e) {
+                throw new \LogicException(\sprintf('Cannot parse interval "%s", please use a valid unit as described on https://www.php.net/datetime.formats.relative.', $interval));
```
Comment (derrabus) [inline src/Symfony/Component/RateLimiter/RateLimiterFactory.php]:
```suggestion
                throw new \LogicException(\sprintf('Cannot parse interval "%s", please use a valid unit as described on https://www.php.net/datetime.formats.relative.', $interval), 0, $e);
```

**Review by derrabus — APPROVED** [review]

**Review by alexandre-daubois — APPROVED** [review]
Are there other places where this could be applicable? Or maybe this is the only place in the code relevant?

**Comment by xabbuh** [comment]
The other calls of `modify()` look safe to me.

**Pull request merged**